``bugprone-bad-signal-to-kill-thread`` is not detected by clangd

The ``bugprone-bad-signal-to-kill-thread`` check is not detected by clangd, but running clang-tidy directly does report the issue.

After some search, I found that clangd only supports matcher-based checks by default, and there’s a ``FastCheckFilter`` option that can be used to force-enable certain checks: https://github.com/clangd/clangd/issues/1727

In the [TidyFastChecks.inc](https://searchfox.org/llvm/source/clang-tools-extra/clangd/TidyFastChecks.inc#54), it is listed as a fast check, so I expected it to be enabled by default.

However, even after setting ``FastCheckFilter: None``, the check still doesn’t seem to be triggered.

Is this the expected behavior, or am I missing something?

**Code**
```c
#include <pthread.h>
#include <signal.h>
#include <unistd.h>

int main(void) {
  pthread_t thread;

  pthread_kill(thread, SIGINT);
  pthread_kill(thread, SIGKILL);
  pthread_kill(thread, SIGTERM);
  pthread_kill(thread, 0xF);

  return 0;
}

```

**.clang-tidy**
```yaml
Checks: '
  -*,
  bugprone-bad-signal-to-kill-thread
  '
```

**.clangd**
```yaml
Index:
  Background: Build

InlayHints:
  Enabled: Yes
  Designators: Yes
  ParameterNames: Yes
  DeducedTypes: Yes

Diagnostics:
  ClangTidy:
    FastCheckFilter: None
```

**settings.json**
```json
    "clangd.arguments": [
        "--all-scopes-completion",
        "--background-index",
        "--background-index-priority=background",
        "--clang-tidy",
        "--compile-commands-dir=${workspaceFolder}/.vscode",
        "--completion-style=detailed",
        "--completion-parse=auto",
        "--enable-config",
        "--fallback-style=gnu",
        "--function-arg-placeholders=true",
        "--header-insertion=never",
        "--header-insertion-decorators",
        "--log=verbose",
        "--pch-storage=memory",
        "--pretty",
        "--ranking-model=decision_forest",
        "-j=6"
    ]
```

**Logs**
```
> clang-tidy --config-file=.clang-tidy -p .vscode/ ./src.c

2 warnings generated.
src.c:10:3: warning: thread should not be terminated by raising the 'SIGTERM' signal [bugprone-bad-signal-to-kill-thread]
   10 |   pthread_kill(thread, SIGTERM);
      |   ^
src.c:11:3: warning: thread should not be terminated by raising the 'SIGTERM' signal [bugprone-bad-signal-to-kill-thread]
   11 |   pthread_kill(thread, 0xF);
      |   ^
```

clangd output: https://gist.github.com/Carryme1899/daab9e124a9ef81d2d7ab78a7cd00b5b

**System information**
Clangd version: clangd version 20.1.8 (Fedora 20.1.8-3.fc42)
clangd extension version: 0.2.0
Operating system: Fedora 42

--- a/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
@@ -42,12 +42,11 @@ void BadSignalToKillThreadCheck::check(const MatchFinder::MatchResult &Result) {
     if (!T.isLiteral())
       return std::nullopt;
 
-    const char *Buffer = nullptr;
+    SmallVector<char> Buffer;
     bool Invalid = false;
-    const unsigned Length = PP->getSpelling(T, Buffer, &Invalid);
+    StringRef ValueStr = PP->getSpelling(T, Buffer, &Invalid);
     if (Invalid)
       return std::nullopt;
-    StringRef ValueStr(Buffer, Length);
 
     llvm::APInt IntValue;
     constexpr unsigned AutoSenseRadix = 0;

--- a/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
@@ -33,14 +33,32 @@ void BadSignalToKillThreadCheck::check(const MatchFinder::MatchResult &Result) {
            KeyValue.first->hasMacroDefinition();
   };
   const auto TryExpandAsInteger =
-      [](Preprocessor::macro_iterator It) -> std::optional<unsigned> {
+      [&Result](Preprocessor::macro_iterator It) -> std::optional<unsigned> {
     if (It == PP->macro_end())
       return std::nullopt;
     const MacroInfo *MI = PP->getMacroInfo(It->first);
     const Token &T = MI->tokens().back();
-    if (!T.isLiteral() || !T.getLiteralData())
+
+    if (!T.isLiteral())
       return std::nullopt;
-    const StringRef ValueStr = StringRef(T.getLiteralData(), T.getLength());
+
+    StringRef ValueStr;
+    if (T.getLiteralData()) {
+      ValueStr = StringRef(T.getLiteralData(), T.getLength());
+    } else {
+      const SourceManager *SM = Result.SourceManager;
+      const SourceLocation Loc = T.getLocation();
+      if (Loc.isInvalid())
+        return std::nullopt;
+      std::optional<StringRef> Buffer =
+          SM->getBufferDataOrNone(SM->getFileID(Loc));
+      if (!Buffer)
+        return std::nullopt;
+      const unsigned Offset = SM->getFileOffset(Loc);
+      if (Offset + T.getLength() > Buffer->size())
+        return std::nullopt;
+      ValueStr = Buffer->substr(Offset, T.getLength());
+    }
 
     llvm::APInt IntValue;
     constexpr unsigned AutoSenseRadix = 0;

--- a/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
@@ -44,7 +44,7 @@ void BadSignalToKillThreadCheck::check(const MatchFinder::MatchResult &Result) {
 
     SmallVector<char> Buffer;
     bool Invalid = false;
-    StringRef ValueStr = PP->getSpelling(T, Buffer, &Invalid);
+    const StringRef ValueStr = PP->getSpelling(T, Buffer, &Invalid);
     if (Invalid)
       return std::nullopt;
 

--- a/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/BadSignalToKillThreadCheck.cpp
@@ -33,7 +33,7 @@ void BadSignalToKillThreadCheck::check(const MatchFinder::MatchResult &Result) {
            KeyValue.first->hasMacroDefinition();
   };
   const auto TryExpandAsInteger =
-      [&Result](Preprocessor::macro_iterator It) -> std::optional<unsigned> {
+      [](Preprocessor::macro_iterator It) -> std::optional<unsigned> {
     if (It == PP->macro_end())
       return std::nullopt;
     const MacroInfo *MI = PP->getMacroInfo(It->first);
@@ -42,23 +42,12 @@ void BadSignalToKillThreadCheck::check(const MatchFinder::MatchResult &Result) {
     if (!T.isLiteral())
       return std::nullopt;
 
-    StringRef ValueStr;
-    if (T.getLiteralData()) {
-      ValueStr = StringRef(T.getLiteralData(), T.getLength());
-    } else {
-      const SourceManager *SM = Result.SourceManager;
-      const SourceLocation Loc = T.getLocation();
-      if (Loc.isInvalid())
-        return std::nullopt;
-      std::optional<StringRef> Buffer =
-          SM->getBufferDataOrNone(SM->getFileID(Loc));
-      if (!Buffer)
-        return std::nullopt;
-      const unsigned Offset = SM->getFileOffset(Loc);
-      if (Offset + T.getLength() > Buffer->size())
-        return std::nullopt;
-      ValueStr = Buffer->substr(Offset, T.getLength());
-    }
+    const char *Buffer = nullptr;
+    bool Invalid = false;
+    const unsigned Length = PP->getSpelling(T, Buffer, &Invalid);
+    if (Invalid)
+      return std::nullopt;
+    StringRef ValueStr(Buffer, Length);
 
     llvm::APInt IntValue;
     constexpr unsigned AutoSenseRadix = 0;

--- a/clang-tools-extra/clangd/unittests/DiagnosticsTests.cpp
+++ b/clang-tools-extra/clangd/unittests/DiagnosticsTests.cpp
@@ -823,6 +823,25 @@ TEST(DiagnosticTest, ClangTidyNoLiteralDataInMacroToken) {
   EXPECT_THAT(TU.build().getDiagnostics(), UnorderedElementsAre()); // no-crash
 }
 
+TEST(DiagnosticTest, BadSignalToKillThreadInPreamble) {
+  Annotations Main(R"cpp(
+    #include "signal.h"
+    using pthread_t = int;
+    int pthread_kill(pthread_t thread, int sig);
+    int func() {
+      pthread_t thread;
+      return pthread_kill(thread, 15);
+    }
+  )cpp");
+  TestTU TU = TestTU::withCode(Main.code());
+  TU.HeaderFilename = "signal.h";
+  TU.HeaderCode = "#define SIGTERM 15";
+  TU.ClangTidyProvider = addTidyChecks("bugprone-bad-signal-to-kill-thread");
+  EXPECT_THAT(TU.build().getDiagnostics(),
+              ifTidyChecks(UnorderedElementsAre(
+                  diagName("bugprone-bad-signal-to-kill-thread"))));
+}
+
 TEST(DiagnosticTest, ClangTidyMacroToEnumCheck) {
   Annotations Main(R"cpp(
     #if 1

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -144,6 +144,10 @@ Changes in existing checks
   <clang-tidy/checks/bugprone/argument-comment>` to also check for C++11
   inherited constructors.
 
+- Improved :doc:`bugprone-bad-signal-to-kill-thread
+  <clang-tidy/checks/bugprone/bad-signal-to-kill-thread>` check by fixing false
+  negatives when the ``SIGTERM`` macro is obtained from a precompiled header.
+
 - Improved :doc:`bugprone-exception-escape
   <clang-tidy/checks/bugprone/exception-escape>` check by adding
   `TreatFunctionsWithoutSpecificationAsThrowing` option to support reporting

